lua/luasnip/loaders/from_lua.lua

```diff
@@ -0,0 +1,209 @@
+-- loads snippets from directory structured almost like snipmate-collection:
+-- - files all named <ft>.lua
+-- - each returns table containing keys (optional) "snippets" and
+--   "autosnippets", value for each a list of snippets.
+
+local cache = require("luasnip.loaders._caches").lua
+local path_mod = require("luasnip.util.path")
+local loader_util = require("luasnip.loaders.util")
+local util = require("luasnip.util.util")
+
+local M = {}
+
+local function load_files(ft, files)
+	local ft_snippets = {}
+	local ft_autosnippets = {}
+
+	for _, file in ipairs(files) do
+		-- 0444 = 292, eg. open with rrr.
+		local fd = vim.loop.fs_open(file, "r", 292)
+
+		if not fd then
+			error("Couldn't find file " .. file)
+		end
+
+		local size = vim.loop.fs_fstat(fd).size
+		local func_string = vim.loop.fs_read(fd, size)
+		-- bring snippet-constructors into global scope for that function.
+		func_string = 'require("luasnip").setup_snip_env() ' .. func_string
+		local file_snippets, file_autosnippets = loadstring(func_string)()
+
+		-- make sure these aren't nil.
+		file_snippets = file_snippets or {}
+		file_autosnippets = file_autosnippets or {}
+
+		-- keep track of snippet-source.
+		cache.path_snippets[file] = {
+			snippets = file_snippets,
+			autosnippets = file_autosnippets,
+		}
+
+		vim.list_extend(ft_snippets, file_snippets)
+		vim.list_extend(ft_autosnippets, file_autosnippets)
+
+		-- use lua autocommands here as soon as they're stable.
+		-- stylua: ignore
+		vim.cmd(string.format(
+			[[
+				augroup luasnip_watch_%s
+				autocmd!
+				autocmd BufWritePost %s lua require("luasnip.loaders.from_lua").reload_file("%s")
+			]],
+			-- augroup name may not contain spaces.
+			file:gsub(" ", "_"),
+			-- escape for autocmd-pattern.
+			file:gsub(" ", "\\ "),
+			file
+		))
+	end
+
+	ls.add_snippets(ft, ft_snippets, { type = "snippets" })
```

Comment:
here, `ls` has not been defined

Reply:
Ouh, great catch, it's global in my config, so I didn't notice it.
TY :+1: